File: Data Structures & Algorithms/coin-change/submission-5.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int>dp(amount+1,-1);
        dp[0]=0;
        for(int i=1;i<=amount;++i){
            int curr_coin =INT_MAX;
            for(int j=0;j<coins.size();++j){
                if(coins[j]<=i and dp[i-coins[j]]!=-1){
                    curr_coin = min(curr_coin, 1+dp[i-coins[j]]);
                }
            }
            if(curr_coin!=INT_MAX) dp[i]=curr_coin;
            // cout<<i<<" "<<dp[i]<<endl;
        }
        return dp[amount];
    }
};

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-5.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        if len(cost) == 2:
            return min(cost[1],cost[0])

        n = len(cost)
        dp = [0 for i in range(n+1)]
        # dp[0], dp[1] = cost[0], min(cost[0],cost[1])

        for i in range(2,n+1):
            dp[i] = min(cost[i-1]+dp[i-1], cost[i-2]+dp[i-2])
        
        return dp[n]